### PR TITLE
Replacing the HAL fence with a semaphore modeled on Vulkan timeline semaphores

### PR DESCRIPTION
**Pull request by benvanik**
This gives us both compiler and runtime versions of the `hal.semaphore` and provides a basic implementation of the host semaphore (for VMLA/LLVMJIT/etc) and the Vulkan native timeline semaphore (requires the layer, extension, or Vulkan 1.2).

This change intentionally does not remove `hal.ex.submit_and_wait` as that is currently handling resource management. Future changes will add a resource timeline and use the semaphores to advance it as a way to collect resources when no longer required.

### CODE REVIEW

**Review by ScottTodd — COMMENTED** [review]
Still making my way through this. Publishing my comments so far. Generally looks great :D

File: iree/hal/host/host_semaphore_test.cc [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+// Tests that a semaphore that is unused properly cleans itself up.
+TEST(HostSemaphoreTest, NoOp) {
+  HostSemaphore semaphore(123u);
+  ASSERT_OK_AND_ASSIGN(uint64_t value, semaphore.Query());
+  EXPECT_EQ(123u, value);
+}
+
+// Tests that a semaphore will accept new values as it is signaled.
+TEST(HostSemaphoreTest, NormalSignaling) {
```
Comment (ScottTodd) [inline iree/hal/host/host_semaphore_test.cc:37]: Nice tests! Would be nice to pull some of this into the HAL CTS somehow.

File: iree/hal/host/host_submission_queue.cc [hunk 2]
```diff
@@ -176,113 +80,103 @@ Status HostSubmissionQueue::ProcessBatches(ExecuteFn execute_fn) {
...
-  submission->pending_batches.clear();
-
-  // Signal the fence.
-  auto* fence = static_cast<HostFence*>(submission->fence.first);
   if (status.ok()) {
-    RETURN_IF_ERROR(fence->Signal(submission->fence.second));
+    if (!submission->pending_batches.empty()) {
+      // Completed with work remaining? Cause a failure.
+      status = FailedPreconditionErrorBuilder(IREE_LOC)
+               << "Submission ended prior to completion of all batches";
```
Comment (ScottTodd) [inline iree/hal/host/host_submission_queue.cc:159]: Looks like this status isn't actually propagated.

How about
```
if (status.ok() && !submission->pending_batches.empty()) {
  // Completed with work remaining? Cause a failure.
  status = FailedPreconditionErrorBuilder(IREE_LOC)
             << "Submission ended prior to completion of all batches";
}
if (!status.ok()) {
  // Fail all pending ...
  ...
}

list_.take(submission).reset();
```

File: iree/hal/host/host_submission_queue.h [hunk 3]
```diff
@@ -125,20 +77,21 @@ class HostSubmissionQueue {
   };
   struct Submission : public IntrusiveLinkBase<void> {
     absl::InlinedVector<PendingBatch, 4> pending_batches;
-    FenceValue fence;
   };
 
   // Returns true if all wait semaphores in the |batch| are signaled.
-  bool IsBatchReady(const PendingBatch& batch) const;
+  // Returns any failure that status set on any of the wait semaphores.
```
Comment (ScottTodd) [inline iree/hal/host/host_submission_queue.h:83]: Having trouble parsing this. How about rewording?

`Returns if the batch is ready, or returns any failure status set on any of the wait semaphores."

File: iree/hal/semaphore.h [hunk 4]
```diff
@@ -15,45 +15,88 @@
...
 
-// Timeline semaphores act as a fence along a per-semaphore timeline where
-// signaling is done by setting the payload to a monotonically increasing
-// 64-bit integer and waiting is done by blocking until the payload is set
-// greater-than or equal-to the specified value. Timeline semaphores may be
-// waited on or signaled in any order and can be significantly more
-// efficient due to system-level coalescing.
-class TimelineSemaphore : public Semaphore {
+// Synchronization mechanism for host->device, device->host, and host->host,
+// and device->device notification. Semaphores behave like Vulkan timeline
```
Comment (ScottTodd) [inline iree/hal/semaphore.h:38]: nit: two "and"s here also

File: iree/hal/semaphore.h [hunk 5]
```diff
@@ -15,45 +15,88 @@
...
+
+  // Signals the semaphore with a failure. The |status| will be returned from
+  // Query and Signal for the lifetime of the semaphore.
+  virtual void Fail(Status status) = 0;
+
+  // Blocks the caller until the semaphore reaches or exceedes the specified
+  // payload value or the |deadline| elapses.
+  //
+  // Returns success if the wait is successful and the semaphore has at met or
+  // exceeded the required payload value.
```
Comment (ScottTodd) [inline iree/hal/semaphore.h:91]: "has met or exceeded"

File: iree/hal/vulkan/native_timeline_semaphore.cc [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  signal_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_SIGNAL_INFO;
+  signal_info.pNext = nullptr;
+  signal_info.semaphore = handle_;
+  signal_info.value = UINT64_MAX;
+  // NOTE: we don't care about the result in case of failures as we are
+  // failing and the caller will likely be tearing everything down anyway.
+  logical_device_->syms()->vkSignalSemaphore(*logical_device_, &signal_info);
+}
+
+Status NativeTimelineSemaphore::Wait(uint64_t value, absl::Time deadline) {
```
Comment (ScottTodd) [inline iree/hal/vulkan/native_timeline_semaphore.cc:78]: I see that some of this logic is duplicated in `VulkanDevice::WaitSemaphores`. If that logic was moved into `NativeTimelineSemaphore` here, then an `EmulatedTimelineSemaphore` (not using the extension layer) could be slotted in more easily. WDYT?

File: iree/hal/vulkan/vulkan_device.cc [hunk 7]
```diff
@@ -642,60 +642,89 @@ StatusOr<ref_ptr<Event>> VulkanDevice::CreateEvent() {
...
-  IREE_TRACE_SCOPE0("VulkanDevice::CreateBinarySemaphore");
+StatusOr<ref_ptr<Semaphore>> VulkanDevice::CreateSemaphore(
+    uint64_t initial_value) {
+  IREE_TRACE_SCOPE0("VulkanDevice::CreateSemaphore");
+
+  VkSemaphoreTypeCreateInfo timeline_create_info;
+  timeline_create_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_TYPE_CREATE_INFO;
+  timeline_create_info.pNext = nullptr;
+  timeline_create_info.semaphoreType = VK_SEMAPHORE_TYPE_TIMELINE;
+  timeline_create_info.initialValue = initial_value;
```
Comment (ScottTodd) [inline iree/hal/vulkan/vulkan_device.cc:653]: Relating to my prior comment about `NativeTimelineSemaphore` / `EmulatedTimelineSemaphore`, this could be moved e.g. `static StatusOr<NativeTimelineSemaphore> NativeTimelineSemaphore::CreateSemaphore(args...)`

(Separation of concerns + more flexibility in emulating if needed, though that *might* be YAGNI... hoping to test on Android soon to see if the extension layer will be sufficient)

File: iree/compiler/Dialect/HAL/IR/HALOps.td [hunk 8]
```diff
@@ -2032,6 +2026,115 @@ def HAL_ExecutableLayoutLookupOp : HAL_PureOp<"executable_layout.lookup", [
...
+    HAL_TimelineValue:$initial_value
+  );
+  let results = (outs
+    HAL_Semaphore:$result
+  );
+
+  let skipDefaultBuilders = 1;
+  let builders = [
+    OpBuilder<[{
+      Builder *builder, OperationState &state, Value device,
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/IR/HALOps.td:2048]: IR types are verified at compile time by MLIR - so device must be HAL_Device. At runtime the values are checked both for being non-null and of the correct type.

File: iree/compiler/Dialect/HAL/IR/HALOps.td [hunk 9]
```diff
@@ -2032,6 +2026,115 @@ def HAL_ExecutableLayoutLookupOp : HAL_PureOp<"executable_layout.lookup", [
...
+      int32_t initial_value
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/IR/HALOps.td:2049]: Unfortunately the HAL_ defs don't end up in the generated C++ types and are only usable by tablegen. It'd be cool if they generated typedefs. I'm trying to use index here for everything (which in MLIR is int64_t), so I'll use that here.

File: iree/compiler/Dialect/HAL/IR/HALOps.td [hunk 10]
```diff
@@ -2032,6 +2026,115 @@ def HAL_ExecutableLayoutLookupOp : HAL_PureOp<"executable_layout.lookup", [
...
+  let assemblyFormat = [{
+    $semaphore attr-dict-with-keyword `:` type($status) `,` type($value)
+  }];
+}
+
+def HAL_SemaphoreSignalOp : HAL_Op<"semaphore.signal"> {
+  let summary = [{semaphore payload value signal operation}];
+  let description = [{
+    Signals the semaphore to the given payload value.
+    The call is ignored if the current payload value exceeds `new_value`.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/IR/HALOps.td:2086]: On our side yes because the values must increase to be signaled signaling to an earlier value should be a no-op. I don't think Vulkan will like that though, so we may need to sanitize things by keeping track of the last value we signaled to and doing the no-op ourselves. It may be a thing we only enable when we are running untrusted models though.

File: iree/compiler/Dialect/HAL/IR/HALOps.td [hunk 11]
```diff
@@ -2032,6 +2026,115 @@ def HAL_ExecutableLayoutLookupOp : HAL_PureOp<"executable_layout.lookup", [
...
+
+  let assemblyFormat = [{
+    $semaphore `,` `status` `=` $status attr-dict-with-keyword
+  }];
+}
+
+// TODO(benvanik): rework this to make it a terminator with branch targets
+// for ^success and ^fail(status).
+def HAL_SemaphoreAwaitOp : HAL_Op<"semaphore.await", [YieldPoint]> {
+  let summary = [{semaphore payload value signal operation}];
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/IR/HALOps.td:2120]: done

File: iree/compiler/Dialect/HAL/hal.imports.mlir [hunk 12]
```diff
@@ -403,4 +403,50 @@ vm.import @executable_layout.create(
...
 attributes {nosideeffects}
 
+//===----------------------------------------------------------------------===//
+// iree::hal::Semaphore
+//===----------------------------------------------------------------------===//
+
+// Returns a semaphore from the device pool with the given initial value.
+vm.import @semaphore.create(
+  %device : !vm.ref<!hal.device>,
+  %initial_value : i32
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/hal.imports.mlir:413]: Here it's just a storage type - it's really signless. I'm not actually sure since Lei's changes for signed types landed whether it's possible to define si32/ui32 or something. Will have to play with that (and fix up everything) at some point.

File: iree/hal/api.cc [hunk 13]
```diff
@@ -1031,6 +1031,111 @@ iree_hal_device_id(iree_hal_device_t* device) {
...
+          src_batch.signal_semaphores.payload_values[j]};
+    }
+    dst_batch.signal_semaphores =
+        absl::MakeConstSpan(&semaphore_values[base_semaphore_index],
+                            src_batch.signal_semaphores.count);
+    base_semaphore_index += src_batch.signal_semaphores.count;
+  }
+
+  // For now we always go to the first compute queue. TBD cleanup pending the
+  // device modeling in the IR as to how we really want to handle this.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/api.cc:1083]: done

File: iree/hal/api.h [hunk 14]
```diff
@@ -966,6 +1014,55 @@ iree_hal_device_allocator(iree_hal_device_t* device);
...
+// semaphores may have been signaled and each can be queried to see which ones.
+IREE_API_EXPORT iree_status_t IREE_API_CALL
+iree_hal_device_wait_semaphores_with_deadline(
+    iree_hal_device_t* device, iree_hal_wait_mode_t wait_mode,
+    const iree_hal_semaphore_list_t* semaphore_list, iree_time_t deadline_ns);
+
+// Blocks the caller until the semaphores reach or exceed the specified payload
+// values or the |timeout_ns| elapses.
+// A relative-time version of iree_hal_device_wait_semaphores_with_deadline
+// using the relative nanoseconds from the time the call is made.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/api.h:1059]: It's fine to to not match the IREE timer - since the time is a standard unix wall-clock time it's possible to generate these values in other environments (python, java, etc) in a way that survives queuing/IPC/etc for when it ends up on this side. So if a user wants to provide relative times they use the timeout one and if they want deadlines they get the time based on whatever they want.

File: iree/hal/device.h [hunk 15]
```diff
@@ -33,6 +33,9 @@
 #include "iree/hal/executable_layout.h"
 #include "iree/hal/semaphore.h"
 
+// Win32 macro name conflict:
+#undef CreateSemaphore
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/device.h:37]: added #if for readability - can't put in platform_headers as then there's an include order issue with vulkan headers/etc that will redefine it

File: iree/hal/device_manager.h [hunk 16]
```diff
@@ -153,30 +153,27 @@ class DeviceManager final {
...
-  // If a sequence of submissions are performed then the semaphore relationships
-  // can be used to elide waits. Submit(A)+Submit(B, fence) where there is a
-  // dependency from A->B is safe.
+  // If signal_semaphores are provided they will be signaled when their
+  // corresponding submission has completed. If a sequence of submissions are
+  // performed then the semaphore value relationships can be used to elide
+  // waits.
   //
-  // All provided resources must remain alive until the provided |fence|
+  // All provided resources must remain alive until the provided |semaphore|
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/device_manager.h:161]: each SubmissionBatch has a signal_semaphores - clarified

File: iree/hal/device_manager.h [hunk 17]
```diff
@@ -187,7 +184,7 @@ class DeviceManager final {
   Status Flush();
 
   // Blocks until all outstanding requests have been completed.
-  // This is equivalent to having waited on all outstanding fences.
+  // This is equivalent to having waited on all outstanding semaphores.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/device_manager.h:187]:
```
  // This is equivalent to having waited on all outstanding semaphore signal
  // operations in all previously submitted batches.
```

File: iree/hal/host/host_semaphore_test.cc (hunk 1 above)
Comment (benvanik) [reply iree/hal/host/host_semaphore_test.cc:37]: yeah should be possible - everything here should be valid on all targets, and it'd be nice to have some threading tests too (for host->host waits).

File: iree/hal/host/host_submission_queue.cc (hunk 2 above)
Comment (benvanik) [reply iree/hal/host/host_submission_queue.cc:159]: good catch!

File: iree/hal/vulkan/native_timeline_semaphore.cc (hunk 6 above)
Comment (benvanik) [reply iree/hal/vulkan/native_timeline_semaphore.cc:78]: Single-wait and multi-wait will sometimes have different implementations (though don't in vulkan), but usually the multi-wait is the load-bearing one. The important thing there is that there may be multiple kinds of semaphores passed in: for example, a host semaphore + a device semaphore + imported fd semaphore all being waited on together. I think the real fix here for when we have that heterogeneous semaphore situation is using WaitHandle (or really a derivative of it, as it's pretty old at this point), which was designed to allow this mixed case.

For now I'll just make these single-wait methods call the multi-wait to reduce duplication, but it'll require some special casing for other types (like until we have a mixed use mode just call static NativeTimelineSemaphore::MultiWait, etc). Can do that when we get an emulated timeline semaphore :)

Comment (benvanik) [reply iree/hal/vulkan/native_timeline_semaphore.cc:78]: Nevermind, I forgot the layering: this intentionally doesn't have a Device handle so it can't call the multi-wait method. Will just leave as-is for now.

File: iree/hal/vulkan/vulkan_device.cc (hunk 7 above)
Comment (benvanik) [reply iree/hal/vulkan/vulkan_device.cc:653]: Good thinking! Done.

File: iree/hal/api.cc [hunk 18]
```diff
@@ -1080,8 +1080,12 @@ IREE_API_EXPORT iree_status_t IREE_API_CALL iree_hal_device_queue_submit(
   }
 
   // For now we always go to the first compute queue. TBD cleanup pending the
```
Comment (ScottTodd) [inline iree/hal/api.cc:1082]: nit: outdated comment - no longer always going to the first compute queue after the simple `queue_affinity` handling.

**Review by ScottTodd — APPROVED** [review]
Nice :D

**Comment by ScottTodd** [comment]
Still working on merging this. There was an issue with SwiftShader crashing when running `iree/hal/cts/allocator_test`, which I've worked around, and now `iree/test/e2e/vulkan_specific/reduce.mlir` has errors running under validation layers that also deadlock when running under SwiftShader.

Logs snippet from `reduce.mlir`:

```
46: I ../iree/tools/run_mlir_main.cc:291] Evaluating all functions in module for driver 'vulkan'...
46: I ../iree/tools/vm_util.cc:304] Creating driver and device for 'vulkan'...
46: E ../iree/hal/vulkan/debug_reporter.cc:40] Cannot call vkDestroySemaphore on VkSemaphore 0xc000000000c[] that is currently in use by a command buffer. The Vulkan spec states: All submitted batches that refer to semaphore must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroySemaphore-semaphore-01137)
46: E ../iree/hal/vulkan/debug_reporter.cc:40] Attempt to free VkCommandBuffer 0x54bad60[] which is in use. The Vulkan spec states: All elements of pCommandBuffers must not be in the pending state (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeCommandBuffers-pCommandBuffers-00047)
46: I ../iree/tools/vm_util.cc:204] result[0]: Buffer<sint32[2]>
46: E ../iree/hal/vulkan/debug_reporter.cc:40] Cannot call vkDestroyPipeline on VkPipeline 0x70000000007[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to pipeline must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyPipeline-pipeline-00765)
```

**Comment by ScottTodd** [comment]
> Still working on merging this. There was an issue with SwiftShader crashing when running `iree/hal/cts/allocator_test`, which I've worked around, and now `iree/test/e2e/vulkan_specific/reduce.mlir` has errors running under validation layers that also deadlock when running under SwiftShader.
> 
> Logs snippet from `reduce.mlir`:
> 
> ```
> 46: I ../iree/tools/run_mlir_main.cc:291] Evaluating all functions in module for driver 'vulkan'...
> 46: I ../iree/tools/vm_util.cc:304] Creating driver and device for 'vulkan'...
> 46: E ../iree/hal/vulkan/debug_reporter.cc:40] Cannot call vkDestroySemaphore on VkSemaphore 0xc000000000c[] that is currently in use by a command buffer. The Vulkan spec states: All submitted batches that refer to semaphore must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroySemaphore-semaphore-01137)
> 46: E ../iree/hal/vulkan/debug_reporter.cc:40] Attempt to free VkCommandBuffer 0x54bad60[] which is in use. The Vulkan spec states: All elements of pCommandBuffers must not be in the pending state (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeCommandBuffers-pCommandBuffers-00047)
> 46: I ../iree/tools/vm_util.cc:204] result[0]: Buffer<sint32[2]>
> 46: E ../iree/hal/vulkan/debug_reporter.cc:40] Cannot call vkDestroyPipeline on VkPipeline 0x70000000007[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to pipeline must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyPipeline-pipeline-00765)
> ```

This was (at least in part) an issue with the validation layers not handling timeline semaphores signalling command buffer completion.
